include/onnxruntime/core/session/onnxruntime_cxx_inline.h

```diff
@@ -288,6 +288,98 @@ inline char* Session::EndProfiling(OrtAllocator* allocator) const {
+  std::vector<Value> output_values;
+  for (size_t i = 0; i < output_node_count; i++) output_values.emplace_back(nullptr);
+  Run(input_values, output_values, run_options);
+  return output_values;
+}
+
+inline void ExperimentalSession::Run(const std::vector<Value>& input_values, std::vector<Value>& output_values, const RunOptions& run_options) {
+  auto input_names = GetInputNames();
+  auto output_names = GetOutputNames();
+  size_t input_node_count = input_names.size();
+  size_t output_node_count = output_names.size();
+  std::vector<const char*> input_names_(input_node_count, nullptr);
+  size_t i = 0;
+  for (auto it=input_names.begin(); it != input_names.end(); it++) input_names_[i++] = (*it).c_str();
+  std::vector<const char*> output_names_(output_node_count, nullptr);
+  i = 0;
+  for (auto it=output_names.begin(); it != output_names.end(); it++) output_names_[i++] = (*it).c_str();
+  Session::Run(run_options, input_names_.data(), input_values.data(), input_node_count, output_names_.data(), output_values.data(), output_node_count);
+}
+
+template <typename Tp> inline
+std::vector<Value> ExperimentalSession::Run(const std::vector<Value>& input_values, const Tp& input_names, const Tp& output_names, const RunOptions& run_options) {
+  static_assert(std::is_same<typename Tp::value_type, std::string>::value, "expect a sequence STL container of objects of type std::string");
+  size_t output_node_count = GetOutputNames().size();
+  std::vector<Value> output_values;
+  for (size_t i = 0; i < output_node_count; i++) output_values.emplace_back(nullptr);
+  Run(input_values, output_values, input_names, output_names, run_options);
+  return output_values;
+}
+
+template <typename Tp> inline
+void ExperimentalSession::Run(const std::vector<Value>& input_values, std::vector<Value>& output_values, const Tp& input_names, const Tp& output_names, const RunOptions& run_options) {
+  static_assert(std::is_same<typename Tp::value_type, std::string>::value, "expect a sequence STL container of objects of type std::string");
+  size_t input_node_count = input_names.size();
+  size_t output_node_count = output_names.size();
+  std::vector<const char*> input_names_(input_node_count, nullptr);
+  size_t i = 0;
+  for (auto it=input_names.begin(); it != input_names.end(); it++) input_names_[i++] = (*it).c_str();
+  std::vector<const char*> output_names_(output_node_count, nullptr);
+  i = 0;
+  for (auto it=output_names.begin(); it != output_names.end(); it++) output_names_[i++] = (*it).c_str();
+  Session::Run(run_options, input_names_.data(), input_values.data(), input_node_count, output_names_.data(), output_values.data(), output_node_count);
+}
+
+inline std::vector<std::string> ExperimentalSession::GetInputNames() const {
+  Ort::AllocatorWithDefaultOptions allocator;
+  size_t node_count = GetInputCount();
+  std::vector<std::string> out(node_count);
+  for (size_t i=0; i<node_count; i++) out[i] = GetInputName(i, allocator);
```

Comment:
This leaks memory, you must free the return value from GetInputName() by using allocator.Free(return value) after it's converted to a std::string

(same applies to GetOutputNames() below

Reply:
I wish we didn't have to query the session, allocate & copy, pass it back down (to where we already have a redundant copy... but refactoring to use the const string names that already exist in the session if we omit the names is a big job (having just poked around down there). This simplified Run case would bypass a lot of code (in retrospect, the way names are mapped to the inputs and outputs seems too complex)...
...anyway I don't like it but I guess we will just have to do it this way if we want to provide a convenient interface today. The needless code is still running, but at least it requires less boilerplate from the end user of the API.

I wonder if there is a fancy way to make GetInputName actually use the standard allocator owned by the string objects we have in the vector, so that the memory is properly owned by those string objects and thus cleaned up automatically when the vector goes out of scope... signing off for the day, didn't get the chance to try

Reply:
Yeah, it's not perfect but it's meant to be an API that takes named parameters. Removing the names now implies an ordering to unnamed parameters which is not currently defined (the order they happen to be indexed by GetInputName & SessionGetOverridableInitializerName is not related to Run).

So for cases where you already know the names (like your own model), it shouldn't be so bad to pass the names in, but that's why the names need to be passed in.

Reply:
Leaks have been addressed in `GetInputNames()`, `GetOutputNames()`, and `GetOverridableInitializerNames()`. Thanks for spotting that!

@RyanUnderhill Thanks for the explanation about the input names. After reading through the comments in this PR, I am starting to understand the need to specify input/output names. In my use cases (computer vision models), nearly all models have a single input and output node so running inference seems like it should be easier...But I'll deal with writing project-specific wrappers for those cases for now.